### PR TITLE
3174 Added tests for AxisAlignedBoxShape

### PR DESCRIPTION
**Pull request by jjjoness**
Signed-off-by: John Jones-Steele <jjjoness@amazon.com>

### CODE REVIEW

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        // 10.0 - 1.0 ~= 9 (box not rotated even though entity is created with rotation in X and Z)
+        EXPECT_TRUE(rayHit);
+        EXPECT_NEAR(distance, 9.00f, 1e-2f);
+    }
+
+   TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess3)
+    {
+        AZ::Entity entity;
+        CreateAxisAlignedBox(
+            AZ::Transform::CreateFromQuaternionAndTranslation(AZ::Quaternion::CreateIdentity(), AZ::Vector3(100.0f, 100.0f, 0.0f)),
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:216]: Nit:  Could use CreateFromTranslation

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+       // transformed scaled
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess4)
+    {
+        AZ::Entity entity;
+        CreateAxisAlignedBox(
+            AZ::Transform::CreateFromQuaternionAndTranslation(
+                AZ::Quaternion::CreateFromAxisAngle(AZ::Vector3::CreateAxisY(), AZ::Constants::QuarterPi), AZ::Vector3(0.0f, 0.0f, 5.0f)) *
+                AZ::Transform::CreateUniformScale(3.0f),
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:236]: Nit: Could just use AZ::Transform(translation, rotation, scale)

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+        EXPECT_TRUE(rayHit);
+        EXPECT_NEAR(distance, 4.0f, 1e-2f);
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxFailure)
+    {
+        AZ::Entity entity;
+        CreateAxisAlignedBox(
+            AZ::Transform::CreateFromQuaternionAndTranslation(AZ::Quaternion::CreateIdentity(), AZ::Vector3(0.0f, -10.0f, 0.0f)),
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:253]: Nit: Could just use CreateFromTranslation

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            }
+        }
+
+        return true;
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, NormalDistributionRandomPointsAreInBox)
+    {
+        // don't rotate transform so that this is an AABB
+        const AZ::Transform transform = AZ::Transform::CreateTranslation(AZ::Vector3(5.0f, 5.0f, 5.0f));
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:127]: Wouldn't this test be more valuable if it verified that things worked when rotated, as opposed to when it isn't?

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        CreateDefaultAxisAlignedBox(transform, entity);
+
+        const bool allRandomPointsInVolume = RandomPointsAreInAxisAlignedBox(entity, AZ::RandomDistributionType::Normal);
+        EXPECT_TRUE(allRandomPointsInVolume);
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, UniformRealDistributionRandomPointsAreInBox)
+    {
+        // don't rotate transform so that this is an AABB
+        const AZ::Transform transform = AZ::Transform::CreateTranslation(AZ::Vector3(5.0f, 5.0f, 5.0f));
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:139]: Same here - Wouldn't this test be more valuable if it verified that things worked when rotated?

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+            AZ::Transform::CreateTranslation(AZ::Vector3(0.0f, 0.0f, 0.0f)) * AZ::Transform::CreateRotationZ(AZ::Constants::QuarterPi),
+            AZ::Vector3(1.0f), entity);
+
+        AZ::Transform transform;
+        AZ::TransformBus::EventResult(transform, entity.GetId(), &AZ::TransformBus::Events::GetWorldTM);
+
+        EXPECT_EQ(transform, AZ::Transform::CreateRotationZ(AZ::Constants::QuarterPi));
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess1)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:173]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        LmbrCentral::ShapeComponentRequestsBus::EventResult(
+            rayHit, entity.GetId(), &LmbrCentral::ShapeComponentRequests::IntersectRay, AZ::Vector3(0.0f, 5.0f, 5.0f),
+            AZ::Vector3(0.0f, -1.0f, 0.0f), distance);
+
+        // 5.0 - 0.5 ~= 4.5 (box not rotated even though entity is created with 45 degree rotation)
+        EXPECT_TRUE(rayHit);
+        EXPECT_NEAR(distance, 4.5f, 1e-2f);
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess2)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:191]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        AZ::TransformBus::EventResult(transform, entity.GetId(), &AZ::TransformBus::Events::GetWorldTM);
+
+        EXPECT_EQ(transform, AZ::Transform::CreateRotationZ(AZ::Constants::QuarterPi));
+    }
+
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess1)
+    {
+        AZ::Entity entity;
+        CreateAxisAlignedBox(
+            AZ::Transform::CreateTranslation(AZ::Vector3(0.0f, 0.0f, 5.0f)) * AZ::Transform::CreateRotationZ(AZ::Constants::QuarterPi),
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:177]: Changed to be Z rotation only.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            AZ::Vector3(1.0f), entity);
+
+        bool rayHit = false;
+        float distance;
+        LmbrCentral::ShapeComponentRequestsBus::EventResult(
+            rayHit, entity.GetId(), &LmbrCentral::ShapeComponentRequests::IntersectRay, AZ::Vector3(0.0f, 5.0f, 5.0f),
+            AZ::Vector3(0.0f, -1.0f, 0.0f), distance);
+
+        // 5.0 - 0.5 ~= 4.5 (box not rotated even though entity is created with 45 degree rotation)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:186]: Changed the comment.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        LmbrCentral::ShapeComponentRequestsBus::EventResult(
+            rayHit, entity.GetId(), &LmbrCentral::ShapeComponentRequests::IntersectRay, AZ::Vector3(-10.0f, -10.0f, 0.0f),
+            AZ::Vector3(0.0f, 0.0f, -1.0f), distance);
+
+        // 10.0 - 1.0 ~= 9 (box not rotated even though entity is created with rotation in X and Z)
+        EXPECT_TRUE(rayHit);
+        EXPECT_NEAR(distance, 9.00f, 1e-2f);
+    }
+
+   TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess3)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:212]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        LmbrCentral::ShapeComponentRequestsBus::EventResult(
+            rayHit, entity.GetId(), &LmbrCentral::ShapeComponentRequests::IntersectRay, AZ::Vector3(100.0f, 100.0f, -100.0f),
+            AZ::Vector3(0.0f, 0.0f, 1.0f), distance);
+
+        EXPECT_TRUE(rayHit);
+        EXPECT_NEAR(distance, 97.5f, 1e-2f);
+    }
+
+       // transformed scaled
+    TEST_F(AxisAlignedBoxShapeTest, GetRayIntersectBoxSuccess4)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:230]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 1 above)
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:216]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 2 above)
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:236]: Done.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 3 above)
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:253]: Done

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 5 above)
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:139]: Changed.

**Review by sphrose — APPROVED** [review]

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        AZ::Entity entity;
+        CreateAxisAlignedBoxWithNonUniformScale(
+            AZ::Transform(
+                AZ::Vector3(2.0f, -5.0f, 3.0f), AZ::Quaternion::CreateFromAxisAngle(AZ::Vector3::CreateAxisY(), AZ::Constants::QuarterPi),
+                0.5f),
+            AZ::Vector3(2.2f, 1.8f, 0.4f), AZ::Vector3(0.2f, 2.6f, 1.2f), entity);
+
+        // This test creates a box of dimensions (2.2, 1.8, 0.4) centred on (2.0, -5, 3) and rotated about the Y axis by 45 degrees.
+        // The box is tested for axis-alignment by firing various rays and ensuring they either hot or miss the box. Any failure hre
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:201]: "hot" -> "hit".  "hre" -> "here"

**Review by mbalfour-amzn — APPROVED** [review]

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp (hunk 12 above)
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:201]: Changed.

File: Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+            AZ::Transform::CreateRotationZ(AZ::Constants::QuarterPi),
+            AZ::Vector3(1.0f), entity);
+
+        bool rayHit = false;
+        float distance;
+        LmbrCentral::ShapeComponentRequestsBus::EventResult(
+            rayHit, entity.GetId(), &LmbrCentral::ShapeComponentRequests::IntersectRay, AZ::Vector3(5.0f, 0.0f, 0.0f),
+            AZ::Vector3(-1.0f, 0.0f, 0.0f), distance);
+
+        // This test creates a unit box centred on (0, 0, 0) and rotated by 45 degrees. The distance to the box should
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp:124]: Oops, English rather than American. Changed throughout the file.

**Pull request merged**